Improve the derivative of the QR decomposition

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #76115

We derive and implement a more concise rule for the forward and backward
derivatives of the QR decomposition. While doing this we:
- Fix the composite compliance of `linalg.qr` and we make it support functorch's batching transform
- Add preconditions to the forward AD.
- Correct / improve the docs, as they contained some inaccuracies.
- Improve the performance and simplify the implementation of both forward and backward
- Avoid saving the input matrix for the backward computation.
- Activate the tests when there's no magma but there's cusolver.

lezcano committed:
Improve derivative of QR decomposition

We derive and implement a more concise rule for the forward and backward
derivatives of the QR decomposition. While doing this we:
- Fix the composite compliance of `linalg.qr` and we make it support batches
- Improve the performance and simplify the implementation of both foward and backward
- Avoid saving the input matrix for the backward computation.

[ghstack-poisoned]

Lezcano commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -3155,170 +3155,171 @@ std::tuple<Tensor, Tensor> linalg_lstsq_backward(
+  }
+  if (m >= n) {
+    const auto syminvadj = [](const Tensor& X) {
+      return X + X.mH() - at::diag_embed(at::real(X.diagonal(0, -2, -1)));
```
@soulitzer  If you change this by, say
```c++
auto ret = X + X.mH();
at::real(ret.diagonal(0,-2, -1)).mul_(0.5);
return ret;
```
the composite compliance test breaks.

soulitzer replied:
Ok the issue seems to be that when we wrap the output in CompositeCompliantTensor, `_make_wrapper_subclass` ignores the strides if it sees that the size is zero and always sets it to 1.

```
elem = torch.ones(0,).as_strided((0,), (2,), 0)
ret = CompositeCompliantTensor(elem)
print(f"{ret.shape = }, {Tensor.stride(ret) = }, {ret.elem.shape = }, {ret.elem.stride() = }")
```

Prints:
```
ret.shape = torch.Size([0]), Tensor.stride(ret) = (1,), ret.elem.shape = torch.Size([0]), ret.elem.stride() = (2,)
```

Either we fix `make_wrapper_subclass` or update CompositeCompliantTensor to skip the check for strides when size is 0.

cc @zou3519

Lezcano replied:
fwiw, once this problem is fixed, I think that a number of `linalg` functions will start being composite compliant. For exmaple the eigenvalue decompositions.

vadimkantorov commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -3155,170 +3155,171 @@ std::tuple<Tensor, Tensor> linalg_lstsq_backward(
+  // and define syminv(X) = triu(X) - 0.5 * diag(X) the inverse of sym : Triu(k) -> Her(k) to give
+  // dR = syminv(sym(Q^H dA R^{-1}))R
+  //
+  // Case m <= n
```
`m < n`?

nikitaved replied:
You mean `m > n`?

vadimkantorov replied:
The other case checks `m >= n` in the code and in the comment, the other case is `else`. So I'm just proposing consistency between the code and the comment and clearness which branch is followed for `m == n`

lezcano committed:
Update on "Improve derivative of QR decomposition"


We derive and implement a more concise rule for the forward and backward
derivatives of the QR decomposition. While doing this we:
- Fix the composite compliance of `linalg.qr` and we make it support batches
- Improve the performance and simplify the implementation of both foward and backward
- Avoid saving the input matrix for the backward computation.

[ghstack-poisoned]

Lezcano commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -3155,170 +3155,177 @@ std::tuple<Tensor, Tensor> linalg_lstsq_backward(
+Tensor linalg_qr_backward(const Tensor& gQ, const Tensor& gR,
+                          const Tensor& Q, const Tensor& R,
+                          const c10::string_view mode) {
+  // Nb. We won't be too formal below, as writing this proof formaly is a pain
+  // We'll link here a formal writing of all this at some point in the future
```
I wrote down the proof of the formulas, in line with most of the other implementations of forward/backward linalg formulas in PyTorch.

The point of this note is that the formulas derived in those links are not formal (even more, the first one refers to another preprint for the derivation). Even more, they just provide the formulas for the case m >= n. I wrote down the derivation for the case m <= n (m < n in the implementation) as well, so this should do for now. The second line of the comment points out that, once the document I'm writing is ready, I'll replace all the derivations with links to this comprehensive paper, which should present all these formulas them in a formal and concise way.

Lezcano replied:
Offline sync. It's good to replace the references by the actual derivation to have the code and the comment explaining the code in sync.

lezcano committed:
Update on "Improve the derivative of the QR decomposition"


We derive and implement a more concise rule for the forward and backward
derivatives of the QR decomposition. While doing this we:
- Fix the composite compliance of `linalg.qr` and we make it support functorch's batching transform
- Add preconditions to the forward AD.
- Improve the performance and simplify the implementation of both foward and backward
- Avoid saving the input matrix for the backward computation.

[ghstack-poisoned]

lezcano committed:
Update on "Improve the derivative of the QR decomposition"


We derive and implement a more concise rule for the forward and backward
derivatives of the QR decomposition. While doing this we:
- Fix the composite compliance of `linalg.qr` and we make it support functorch's batching transform
- Add preconditions to the forward AD.
- Correct / improve the docs, as they contained some inaccuracies.
- Improve the performance and simplify the implementation of both forward and backward
- Avoid saving the input matrix for the backward computation.
- Activate the tests when there's no magma but there's cusolver.

[ghstack-poisoned]

nikitaved commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -11238,13 +11238,13 @@
   variants: function
 
-- func: linalg_qr(Tensor self, str mode='reduced') -> (Tensor Q, Tensor R)
+- func: linalg_qr(Tensor A, str mode='reduced') -> (Tensor Q, Tensor R)
```
So, we prefer a reduced form of QR while SVD for some reason is full :)

Lezcano replied:
I KNOW RIGHT??!!!! -.-

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -3155,170 +3155,177 @@ std::tuple<Tensor, Tensor> linalg_lstsq_backward(
-  return std::get<0>(linalg_qr_jvp(dA, Q, R));
-}
+    const auto sym = [](const Tensor& X) { return X + X.mH(); };
+    const auto syminv = [](const Tensor& X) {
+      auto ret = X.triu();
+      ret.diagonal(0, -2, -1).mul_(0.5);
+      return ret;
+    };
```
Fun to see that the input is const while the function has a side effect :)

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -3155,170 +3155,177 @@ std::tuple<Tensor, Tensor> linalg_lstsq_backward(
+  // Q^H dA R^{-1} = Q^HdQ + dRR^{-1}
+  // where Q^HdQ is skew Hermitian and dRR^{-1} is upper triangular
+  // Define sym(X) = X + X^H
+  // sym(dRR^{-1}) = sym(Q^H dA R^{-1})
+  // and define syminv(X) = triu(X) - 0.5 * diag(X) the inverse of sym : Triu(k) -> Her(k) to give
+  // dR = syminv(sym(Q^H dA R^{-1}))R
```
I think this works because the diagonal of `R` is real?

Lezcano replied:
Indeed. It read: "the inverse of `sym : Triu(k, diag \in R) -> \Her(k)`"

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -3155,170 +3155,177 @@ std::tuple<Tensor, Tensor> linalg_lstsq_backward(
+  // Q^H A_1 R_1^{-1} = Q^H dQ + dR_1 R_1^{-1}
+  // Define trilIm(X) = X.tril(-1) + i * Im diag(X)
+  // trilIm(Q^H dQ) = trilIm(Q^H A_1 R_1^{-1})
+  // and define trilIminv(X) = X - X^H - 0.5*i*Im diag(X). This is the inverse of
```
shouldn't `trillIminv(X) = X - X^H - i * im diag(X)`?

Lezcano replied:
Yup.

nikitaved commented on torch/linalg/__init__.py:
```diff
@@ -2320,20 +2321,17 @@
   This behavior may change in a future PyTorch release.
 
-.. note:: The elements in the diagonal of `R` are not necessarily positive.
+.. note:: The only :attr:`mode` that is differentiable is :attr:`mode`\ `= 'reduced'`.
 
```
Shouldn't `mode=complete` with `m < n` be differentiable as well?

Lezcano replied:
Yeah, the thing is that if `m < n`, mode complete and reduced are equivalent, so I think it's fine not advertising that really. You should be using `reduced`mode anyway.

nikitaved replied:
cc @mruberry , @IvanYashchuk . What is your take on this one?

IvanYashchuk replied:
I'm leaning towards being explicit about what's supported if we started writing this note, i.e. I would include that supported modes are "reduced" or "complete" && `m <= n`.

mruberry replied:
Erring on the side of more detail sounds fine (and maybe an opportunity to suggest a best practice)

lezcano committed:
Update on "Improve the derivative of the QR decomposition"


We derive and implement a more concise rule for the forward and backward
derivatives of the QR decomposition. While doing this we:
- Fix the composite compliance of `linalg.qr` and we make it support functorch's batching transform
- Add preconditions to the forward AD.
- Correct / improve the docs, as they contained some inaccuracies.
- Improve the performance and simplify the implementation of both forward and backward
- Avoid saving the input matrix for the backward computation.
- Activate the tests when there's no magma but there's cusolver.

[ghstack-poisoned]

Lezcano commented:
@soulitzer I run `test_vmapvjpvjp` with the current changes (in-place `at::real(t.diagonal(0, -2, -1)).mul_(0.5);` and it passes. Leaving it like this for now, as it is more efficient.

lezcano committed:
Update on "Improve the derivative of the QR decomposition"


We derive and implement a more concise rule for the forward and backward
derivatives of the QR decomposition. While doing this we:
- Fix the composite compliance of `linalg.qr` and we make it support functorch's batching transform
- Add preconditions to the forward AD.
- Correct / improve the docs, as they contained some inaccuracies.
- Improve the performance and simplify the implementation of both forward and backward
- Avoid saving the input matrix for the backward computation.
- Activate the tests when there's no magma but there's cusolver.

[ghstack-poisoned]

lezcano committed:
Update on "Improve the derivative of the QR decomposition"


We derive and implement a more concise rule for the forward and backward
derivatives of the QR decomposition. While doing this we:
- Fix the composite compliance of `linalg.qr` and we make it support functorch's batching transform
- Add preconditions to the forward AD.
- Correct / improve the docs, as they contained some inaccuracies.
- Improve the performance and simplify the implementation of both forward and backward
- Avoid saving the input matrix for the backward computation.
- Activate the tests when there's no magma but there's cusolver.

[ghstack-poisoned]

Lezcano commented:
@nikitaved this is ready for another review.

nikitaved commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -10672,19 +10672,16 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
+           sample_inputs_func=sample_inputs_linalg_qr_geqrf,
            skips=(
-               DecorateInfo(unittest.expectedFailure, 'TestCompositeCompliance', 'test_backward'),
+               # The test is wrong https://github.com/pytorch/pytorch/pull/76115/files#r854455360
                DecorateInfo(unittest.expectedFailure, 'TestCompositeCompliance', 'test_forward_ad'),
```
Could you please clarify? What is wrong exactly?

Lezcano replied:
They don't handle strides properly in the test for empty tensors.

Lezcano replied:
See the comment linked there for further discussion.

nikitaved replied:
Sorry, I am missing something, but the link is to the PRs hash. Where do I find the comment?

nikitaved replied:
Or do you mean the link that was removed in your PR regarding empty inputs?

nikitaved reviewed (approved):
Thank you, Mario. Looks great!

albanD reviewed (approved):
SGTM

albanD commented:
The lint needs fixing before landing though!

lezcano committed:
Update on "Improve the derivative of the QR decomposition"


We derive and implement a more concise rule for the forward and backward
derivatives of the QR decomposition. While doing this we:
- Fix the composite compliance of `linalg.qr` and we make it support functorch's batching transform
- Add preconditions to the forward AD.
- Correct / improve the docs, as they contained some inaccuracies.
- Improve the performance and simplify the implementation of both forward and backward
- Avoid saving the input matrix for the backward computation.
- Activate the tests when there's no magma but there's cusolver.

[ghstack-poisoned]

lezcano committed:
Update on "Improve the derivative of the QR decomposition"


We derive and implement a more concise rule for the forward and backward
derivatives of the QR decomposition. While doing this we:
- Fix the composite compliance of `linalg.qr` and we make it support functorch's batching transform
- Add preconditions to the forward AD.
- Correct / improve the docs, as they contained some inaccuracies.
- Improve the performance and simplify the implementation of both forward and backward
- Avoid saving the input matrix for the backward computation.
- Activate the tests when there's no magma but there's cusolver.

[ghstack-poisoned]

Lezcano commented:
@pytorchmergebot merge this please
